ReactAndroid/src/main/java/com/facebook/react/views/scroll/ReactHorizontalScrollView.java

```diff
@@ -1239,6 +1288,97 @@ private void updateStateOnScroll() {
     updateStateOnScroll(getScrollX(), getScrollY());
   }
 
+  /**
+   * Called when a mContentView's layout has changed. Fixes the scroll position depending on
+   * maintainVisibleContentPosition
+   */
+  @Override
+  public void onLayoutChange(
+      View v,
+      int left,
+      int top,
+      int right,
+      int bottom,
+      int oldLeft,
+      int oldTop,
+      int oldRight,
+      int oldBottom) {
+    if (mContentView == null) {
+      return;
+    }
+
+    if (this.mMaintainVisibleContentPositionData != null) {
+      scrollMaintainVisibleContentPosition();
+    }
+  }
+
+  /**
+   * Called when maintainVisibleContentPosition is used and after a scroll. Finds the first
+   * completely visible view in the ScrollView and stores it for later use.
+   */
+  private void computeFirstVisibleItemForMaintainVisibleContentPosition() {
+    ReactScrollViewMaintainVisibleContentPositionData maintainVisibleContentPositionData =
+        mMaintainVisibleContentPositionData;
+    if (maintainVisibleContentPositionData == null) return;
+
+    int currentScrollX = getScrollX();
+    int minIdx = maintainVisibleContentPositionData.minIndexForVisible;
+
+    ReactViewGroup contentView = (ReactViewGroup) getChildAt(0);
+    if (contentView == null) return;
+
+    for (int i = minIdx; i < contentView.getChildCount(); i++) {
+      // Find the first entirely visible view. This must be done after we update the content offset
+      // or it will tend to grab rows that were made visible by the shift in position
+      View child = contentView.getChildAt(i);
+      if (child.getX() >= currentScrollX || i == contentView.getChildCount() - 1) {
+        firstVisibleViewForMaintainVisibleContentPosition = new WeakReference<>(child);
+        Rect frame = new Rect();
+        child.getHitRect(frame);
+        prevFirstVisibleFrameForMaintainVisibleContentPosition = frame;
+        break;
+      }
+    }
+  }
+
+  /**
+   * Called when maintainVisibleContentPosition is used and after a layout change. Detects if the
+   * layout change impacts the scroll position and corrects it if needed.
+   */
+  private void scrollMaintainVisibleContentPosition() {
```

Comment:
There are some problems with this approach.

1. When the scroll view first loads and no scroll events occur before items are prepended, this approach would not maintain scroll position because `firstVisibleView` is only set after at least one scroll event.
2. There could be a race condition between when the VirtualizedList prepends an item and when we invoke a scroll event (`onScroll` vs. `onLayoutChange`). This can cause the delta of the cached item to incorrectly be zero, leading to the scroll position not being maintained.
3. We noticed that when there are many items in the list, `onLayoutChange` does not get called and the scroll position is not maintained. (This may be due to the loop computation being too expensive.)

Reply:
> There could be a race condition between when the VirtualizedList prepends an item and when we invoke a scroll event (onScroll vs. onLayoutChange). This can cause the delta of the cached item to incorrectly be zero, leading to the scroll position not being maintained.

I've been digging into this problem and trying to find a solution. Any suggestions?

Reply:
Unless I'm mistaken, the same problem exists with the current iOS implementation of `maintainVisibleContentPosition` and may be the cause of https://github.com/facebook/react-native/issues/25239